extensions/amp-selector/1.0/amp-selector.js

```diff
@@ -0,0 +1,348 @@
+    .filter(
+      (el) =>
+        !closestAncestorElementBySelector(
+          dev().assertElement(el.parentElement),
+          '[option]'
+        )
+    )
+    .forEach((child) => {
+      const option = child.getAttribute('option');
+      const selected = child.hasAttribute('selected');
+      const disabled = child.hasAttribute('disabled');
+      const props = {
+        as: OptionShim,
+        option,
+        // TODO: `disabled` is always undefined for OptionShim
+        isDisabled: disabled,
+        disabled,
+        role: child.getAttribute('role') || 'option',
+        domElement: child,
+        // TODO(wg-bento): This implementation causes infinite loops on DOM mutation.
+        // See https://github.com/ampproject/amp-react-prototype/issues/40.
+        postRender: () => {
+          // Skip mutations to avoid cycles.
+          mu.takeRecords();
+        },
+        selected,
+      };
+      if (selected && option) {
+        value.push(option);
+      }
+      const optionChild = <Option {...props} />;
+      options.push(option);
+      children.push(optionChild);
+    });
+  return dict({'value': value, 'children': children, 'options': options});
+}
+
+/**
+ * Triggers a 'select' event with two data params:
+ * 'targetOption' - option value of the selected or deselected element.
+ * 'selectedOptions' - array of option values of selected elements.
+ *
+ * @param {!Window} win
+ * @param {!../../../src/service/action-impl.ActionService} action
+ * @param {!Element} el The element that was selected or deslected.
+ * @param {string} option
+ * @param {Array<string>} value
+ * @param {!ActionTrust} trust
+ * @private
+ */
+function fireSelectEvent(win, action, el, option, value, trust) {
+  const name = 'select';
+  const selectEvent = createCustomEvent(
+    win,
+    `amp-selector.${name}`,
+    dict({'targetOption': option, 'selectedOptions': value})
+  );
+  action.trigger(el, name, selectEvent, trust);
+}
+
+/**
+ * @param {number} index
+ * @param {!Array<string>} value
+ * @param {Array<string>} options
+ * @param {boolean=} opt_select
+ * @return {JsonObject|undefined}
+ */
+function toggle(index, value, options, opt_select) {
+  userAssert(index, "'index' must be specified");
+  userAssert(index >= 0, "'index' must be greater than 0");
+  userAssert(
+    index < options.length,
+    "'index' must be less than the length of options in the <amp-selector>"
+  );
+  const option = options[index];
+  const target = value.indexOf(option);
+  if (target === -1) {
+    if (opt_select == false) {
+      return;
+    }
+    value.push(option);
+  } else {
+    if (opt_select == true) {
+      return;
+    }
+    value.splice(target, 1);
+  }
+  return dict({'value': value, 'option': option});
+}
+
+/**
+ * @param {number} delta
+ * @param {!Array<string>} value
+ * @param {Array<string>} options
+ * @return {{value: Array<string>, option: string}|undefined}
+ */
+function select(delta, value, options) {
```

Comment:
Can you add a comment on what this function is supposed to do? It's super weird.

Reply:
Done. +1 super weird